contracts/upgradeable_contracts/BaseFeeManager.sol

```diff
@@ -0,0 +1,65 @@
+pragma solidity 0.4.24;
+
+import "../upgradeability/EternalStorage.sol";
+import "../libraries/SafeMath.sol";
+import "../IRewardableValidators.sol";
+
+
+contract BaseFeeManager is EternalStorage {
+    using SafeMath for uint256;
+
+    event FeeUpdated(uint256 fee);
+
+    function calculateFee(uint256 _value, bool _recover) external view returns(uint256) {
+        uint256 fee = getFee();
+        uint256 eth = 1 ether;
+        if (!_recover) {
+            return _value.mul(fee).div(eth);
+        }
+        return _value.mul(fee).div(eth.sub(fee));
+    }
+
+    function setFee(uint256 _fee) external {
+        uintStorage[keccak256(abi.encodePacked("fee"))] = _fee;
+        emit FeeUpdated(_fee);
+    }
+
+    function getFee() public view returns(uint256) {
+        return uintStorage[keccak256(abi.encodePacked("fee"))];
+    }
+
+    function distributeFeeFromAffirmation(uint256 _fee) external {
+        distributeFeeProportionally(_fee, true);
```

Comment:
consider to have two constants REWARD_FOR_TRANSFERRING_FROM_HOME and REWARD_FOR_TRANSFERRING_FROM_FOREIGN. They will be used instead of `true` and `false`. In this case `_isAffirmation` bellow could be replaced by `_direction`.
So, the content of `onFeeDistribution` could look like
```solidity
    function onFeeDistribution(address _rewardAddress, uint256 _fee, bool _direction) internal {
        if (_direction == REWARD_FOR_TRANSFERRING_FROM_FOREIGN) {
            onAffirmationFeeDistribution(_rewardAddress, _fee);
        } else {
            onSignatureFeeDistribution(_rewardAddress, _fee);
        }
    }
```

Reply:
Done https://github.com/poanetwork/poa-bridge-contracts/pull/132/commits/4b8f53a01459a6c31768acad7a8acbe8f079f4ad